### ISSUE TITLE
Dokku Daemon

### ISSUE THREAD

**Issue by josegonzalez** [dokku/dokku#1559]
One thing that is fairly annoying is the inability to orchestrate dokku from afar. Users end up using crappy clients over ssh for many actions, and many have lamented not having a good web interface.

There are a few problems here:

- Log output is very non-standard, and certainly not in a form that can be consumed by an api
- There is no way to safely talk to a shell remotely.

I don't think either of the above is easily solvable in dokku as is, but I do think we can work on allowing people to build such an api.

---

### general spec

- daemon listens on a unix socket
- commands are issued on the cli and responses sent back via streaming json(?)
- commands are validated before being issued. Perhaps via parsing `dokku help`
- no authentication layer.
- one command can be issued at a time to prevent race conditions

### open questions

- How do we send back data to the client and notify them that a command completed? Clients should be aware that it is possible we never send a "terminal" signal.
- We should log commands sent, as well as output, somewhere, if only for debugging purposes.
- How do we make list actions fast? `.3` seconds to list all available plugins is odd.

### why not a full blown api/dashboard

This is the first step into building such a tool. Once this is done, you can then have a separate app running within a docker container that actually contains the api/dashboard. The unix socket can be mounted and the api would talk to it instead of needing to live on the bare host.

Once nice thing about this is that - in theory - we could share a single socket amongst multiple applications, so this would still support running dokku inside of a container.

@Flink @michaelshobbs thoughts?

**Comment by Flink** [comment]
One thing that could be problematic is the validation of the commands. For core functions/plugins I think we won’t run into (too) many problems but with third party plugins nothing can ensure anything. `help` could be not up to date or worse, just wrong.
We could use the `plugin.toml` file to describe every command a plugin provides and their arguments. With that we could easily validate any input. Maybe a similar mechanism could be applied for output, I don’t know, just thinking here :sweat_smile:

**Comment by michaelshobbs** [comment]
Interesting. I see the potential value but I'm not convinced it should be distributed with dokku core. Thoughts there?

**Comment by josegonzalez** [comment]
I'm +1 on it being an official project, and not part of the core.

**Comment by fabdelgado** [comment]
@josegonzalez There is a project for this?

**Comment by josegonzalez** [comment]
Here you go! https://github.com/dokku/dokku-daemon

**Comment by josegonzalez** [comment]
Closing. We have an initial daemon setup thanks to @joshmcrae!
